milestones/api.py

```diff
@@ -0,0 +1,325 @@
+    'relationship': optional filter on milestone relationship type (string, eg: 'requires')
+    'user': optional filter to constrain the set to those milestones which a user has already collected
+    Returns an array of dicts containing milestones
+    """
+    [_validate_course_key(course_key) for course_key in course_keys]  # pylint: disable=expression-not-assigned
+
+    if relationship is not None:
+        _validate_milestone_relationship_type(relationship)
+
+    return data.fetch_courses_milestones(
+        course_keys=course_keys,
+        relationship=relationship,
+        user=user)
+
+
+def remove_course_milestone(course_key, milestone):
+    """
+    Removes the specfied milestone from the specified course
+    """
+    _validate_course_key(course_key)
+    _validate_milestone(milestone)
+    return data.delete_course_milestone(course_key=course_key, milestone=milestone)
+
+
+def add_course_content_milestone(course_key, content_key, relationship, milestone):
+    """
+    Adds a course-content-milestone link to the system
+    'relationship': string value (eg: 'requires')
+    """
+    _validate_course_key(course_key)
+    _validate_content_key(content_key)
+    _validate_milestone_relationship_type(relationship)
+    _validate_milestone(milestone)
+    data.create_course_content_milestone(
+        course_key=course_key,
+        content_key=content_key,
+        relationship=relationship,
+        milestone=milestone)
+
+
+def get_course_content_milestones(course_key, content_key, relationship=None):
+    """
+    Retrieves the set of milestones for a given course content module
+    'relationship': optional filter on milestone relationship type (string, eg: 'requires')
+    Returns an array of dicts containing milestones
+    """
+    _validate_course_key(course_key)
+    _validate_content_key(content_key)
+
+    if relationship is not None:
+        _validate_milestone_relationship_type(relationship)
+
+    return data.fetch_course_content_milestones(
+        course_key=course_key,
+        content_key=content_key,
+        relationship=relationship
+    )
+
+
+def remove_course_content_milestone(course_key, content_key, milestone):
+    """
+    Removes the specified milestone from the specified course content module
+    """
+    _validate_course_key(course_key)
+    _validate_content_key(content_key)
+    _validate_milestone(milestone)
+    return data.delete_course_content_milestone(
+        course_key=course_key,
+        content_key=content_key,
+        milestone=milestone
+    )
+
+
+def add_user_milestone(user, milestone):
+    """
+    Adds a new User-Milestone relationship to the system
+    """
+    _validate_user(user)
+    _validate_milestone(milestone)
+    data.create_user_milestone(user, milestone)
+
+
+def get_user_milestones(user):
+    """
+    Retrieves the set of milestones for a given user
+    Returns an array of dicts
+    """
+    _validate_user(user)
+    return data.fetch_user_milestones(user)
```

Comment:
This could potentially become impractically large, depending on how granular we get. Is there a use case where we'd need to do this, without specifying a namespace those milestones are in?